modules/aws_vpc/main.tf

```diff
@@ -38,44 +44,45 @@ resource "aws_vpc" "vpc" {
 }
 
 resource "aws_vpc_endpoint" "private_endpoints" {
-  count              = length(var.vpc_private_endpoints)
-  vpc_id             = local.vpc_id
-  service_name       = "com.amazonaws.${var.region}.${var.vpc_private_endpoints[count.index]}"
-  vpc_endpoint_type  = "Interface"
-  security_group_ids = [var.security_group_id]
+  for_each            = var.vpc_private_endpoints
+  vpc_id              = local.vpc_id
+  service_name        = "com.amazonaws.${var.region}.${each.key}"
+  vpc_endpoint_type   = each.value
+  security_group_ids  = each.value == "Interface" ? [var.security_group_id] : null
+  private_dns_enabled = each.value == "Interface" ? true : null
 
   tags = merge(
     {
-      "Name" = format("%s", "${var.name}-private-endpoint-${var.vpc_private_endpoints[count.index]}")
+      "Name" = format("%s", "${var.name}-private-endpoint-${each.key}")
     },
     var.tags,
   )
 
-  subnet_ids = [
+  subnet_ids = each.value == "Interface" ? [
     for subnet in local.private_subnets : subnet.id
-  ]
+  ] : null
 }
 
 data "aws_subnet" "public" {
-  count = local.existing_public_subnets ? length(var.subnets["public"]) : 0
+  count = local.existing_public_subnets ? length(var.existing_subnet_ids["public"]) : 0
```

Comment:
They are both arrays of subnets but shouldn't all three of the length expressions contain references to the existing subnet arrays, since that is what we are setting the id's from on the next line in each case?
Getting the length from the default set of subnets, then setting the ids from the existing subnet array would leave us open to an index out of bounds problem, wouldn't it?
It appears that the original logic was incorrectly referring to the length of the "public" array for the default subnets map, instead of the length of the "public" array for the existing subnets map, which is what the data resources created here are referring to. Since the length of those arrays are often the same at "2", either expression would appear to work ok until the lengths of the subnet arrays in those maps are different.

Reply:
Thomas and I discussed the change to the existing_subnet_ids arrays from the default subnet arrays. He was OK to resolve this comment based on our discussion.